### DISCUSSION TITLE
Question about CacheLib's NVM concurrency

### DISCUSSION THREAD

**Discussion by wangyuyue** [facebook/CacheLib#286] [General]
I notice that in both `allocator/nvmcache/NvmCache-inl.h` and `navy/scheduler/` directories, there exists some code related to the handling of concurrent NVM operations for objects of the same key. So what is the difference of responsibilities between them? `allocator/nvmcache/NvmCache-inl.h` is quite complicated, and I really hope somebody can solve my confusion. Thanks a lot.

**Comment by haowu14** [comment]
Hi sorry for the late reply.

If I were to be very specific about the wording:
NvmCache-inl.h is more on concurrency control: making sure that different threads will be able to see a consistent view. For example, lookup would abort on-going puts, etc.

navy/scheduler/ is about parallelism: Given the incoming NVM jobs, how to put them in multiple queues and execute them in parallel.

**Comment by pbhandar2** [comment]
Closing this discussion as it has been inactive. If this is still relevant, then please create a new discussion post and we will get back to you as soon as we can. Thank you!
